lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -225,21 +246,24 @@ class PipelineOutlineLowering : public PipelineLowering {
   // Helper class to manage grabbing the various inputs for stage modules.
   struct PipelineStageMod {
     PipelineStageMod(PipelineOutlineLowering &parent, Block *block,
-                     hw::HWModuleOp mod) {
+                     hw::HWModuleOp mod, bool withStall) {
       size_t nStageInputArgs = parent.pipeline.getNumStageArguments(block);
       inputs = mod.getArguments().take_front(nStageInputArgs);
       llvm::SmallVector<Value> stageExtInputs;
       parent.getStageExtInputs(block, stageExtInputs);
       extInputs =
           mod.getArguments().slice(nStageInputArgs, stageExtInputs.size());
-      valid = mod.getArgument(mod.getNumArguments() - 3);
+      valid = mod.getArgument(mod.getNumArguments() - (withStall ? 4 : 3));
```

Comment:
Please either break these magic numbers into well named constants or include inline comments.

Reply:
Or computing them dynamically by adding the variable length args together. Or by tracking them.

I've had trouble with assuming / recomputing offsets in the past (in the ESI port mapping code), so I refactored it to explicitly track them. Made the code much nicer.